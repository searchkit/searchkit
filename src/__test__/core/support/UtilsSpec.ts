import {Utils} from "../../../"


describe("Utils", ()=> {


  it("guid()", ()=> {
    expect(Utils.guid().length).toEqual(36)
  })

  it("collapse()", ()=> {
    let times2 = n => n*2
    expect(
      Utils.collapse([times2, times2, times2], 1)
    ).toBe(8)
  })

  it("instanceOf", ()=> {
    let isRegex = Utils.instanceOf(RegExp)
    expect(isRegex(/s/)).toBe(true)
    expect(isRegex("s")).toBe(false)
  })

  it("interpolate", ()=> {
    expect(Utils.interpolate("hello {message}, Hello {message}{ending}{missing}", {
      message:"World",
      ending:"!"
    })).toEqual("hello World, Hello World!{missing}")
  })

  it("translate() - don't call interpolate", ()=> {
    let key = "No Results found!"
    spyOn(Utils, "interpolate")
    expect(Utils.translate(key)).toBe(key)
    expect(Utils.interpolate).not.toHaveBeenCalled()
  })

  it("translate() - with interpolations", ()=> {
    let key = "{count} Results found!"
    spyOn(Utils, "interpolate").and.callThrough()
    expect(Utils.translate(key, {count:5})).toBe(
      "5 Results found!"
    )
    expect(Utils.interpolate).toHaveBeenCalledWith(
      key, {count:5}
    )
  })

  it("autobind()", ()=> {
    class Foo {
      message="hello"

      @Utils.autobind
      getMessage(something){
        return this.message + " " + something
      }
    }

    let fn = new Foo().getMessage
    expect(fn("world")).toBe("hello world")

  })
})
