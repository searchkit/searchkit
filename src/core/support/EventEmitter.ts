import {without} from "lodash"
import {each} from "lodash"

export class EventEmitter {
  listeners = []

  addListener(fn){
    this.listeners.push(fn)
    return ()=>{
      this.listeners = without(this.listeners, fn)
    }
  }

  trigger(...args){
    each(this.listeners, (fn)=> {
      fn.apply(null, args)
    })
  }
}
