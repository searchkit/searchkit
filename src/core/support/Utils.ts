import {reduce} from "lodash"
import {map} from "lodash"
import {reject} from "lodash"
import {isUndefined} from "lodash"

export class Utils {
  static guidCounter = 0

  static guid(prefix=""){
    let id = ++Utils.guidCounter
    return prefix.toString() + id
  }

  static collapse(collection, seed){
    const reducer = (current, fn)=> fn(current)
    return reduce(collection, reducer, seed)
  }

  static instanceOf(klass){
    return (val)=> val instanceof klass
  }

  static interpolate(str, interpolations){
    return str.replace(
  		/{([^{}]*)}/g,
  		(a, b) => {
  			var r = interpolations[b];
  			return typeof r === 'string' || typeof r === 'number' ? r : a;
  		}
    )
  }

  static translate(key, interpolations?){
    if(interpolations){
      return Utils.interpolate(key, interpolations)
    } else {
      return key
    }
  }

  static computeOptionKeys(options, fields, defaultKey){
    return map(options, (option)=> {
      return Utils.generateKeyFromFields(option, fields, defaultKey)
    })
  }

  static generateKeyFromFields(ob, fields, defaultKey){
    if(ob.key){
      return ob
    }
    let fieldValues = reject(map(fields, (field: string)=> ob[field]), isUndefined)
    if(fieldValues.length > 0){
      ob.key = fieldValues.join("_")
    } else {
      ob.key = defaultKey
    }
    return ob
  }
}
