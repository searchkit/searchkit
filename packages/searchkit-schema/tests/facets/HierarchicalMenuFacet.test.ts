import nock from 'nock'
import { SearchkitConfig } from '@searchkit/sdk'
import { MultiMatchQuery, HierarchicalMenuFacet } from '@searchkit/sdk'
import { setupTestServer, callQuery } from '../support/helper'
import lvl1Mock from '../__mock-data__/HierarchicalMenuFacet/lvl1.json'
import lvl2Mock from '../__mock-data__/HierarchicalMenuFacet/lvl2.json'
import lvl3Mock from '../__mock-data__/HierarchicalMenuFacet/lvl3.json'

describe('HierarchicalMenuFacet', () => {
  describe('should return as expected', () => {
    const runQuery = async (gql) => {
      const response = await callQuery({ gql })
      return response
    }

    const config: SearchkitConfig = {
      host: 'http://localhost:9200',
      index: 'movies',
      hits: {
        fields: ['actors', 'writers']
      },
      query: new MultiMatchQuery({ fields: ['actors', 'writers', 'title^4', 'plot'] }),
      facets: [
        new HierarchicalMenuFacet({
          fields: ['category_lvl1.keyword', 'category_lvl2.keyword', 'category_lvl3.keyword'],
          identifier: 'categories',
          label: 'Categories'
        })
      ]
    }

    it('No filters, bring back only level 1 facets', async () => {
      setupTestServer({
        config,
        addToQueryType: true,
        typeName: 'ResultSet',
        hitTypeName: 'ResultHit'
      })

      const gql = `
        {
          results(filters: []) {
            hits(page: {size: 10, from: 0 }) {
              items {
                id
              }
            }
            facets {
              identifier
              display
              label
              type
              display
              ... on HierarchicalMenuFacet {
                entries {
                  count
                  label
                  entries {
                    count
                    label
                    entries {
                      count
                      label
                    }
                  }
                }
              }
            }
          }
        }
      `

      const scope = nock('http://localhost:9200')
        .post('/movies/_search')
        .reply((uri, body) => {
          expect(body).toMatchInlineSnapshot(`
            Object {
              "_source": Object {
                "includes": Array [
                  "actors",
                  "writers",
                ],
              },
              "aggs": Object {
                "facet_bucket_all": Object {
                  "aggs": Object {
                    "categories": Object {
                      "aggs": Object {
                        "lvl_1": Object {
                          "aggs": Object {
                            "aggs": Object {
                              "terms": Object {
                                "field": "category_lvl1.keyword",
                              },
                            },
                          },
                          "filter": Object {
                            "match_all": Object {},
                          },
                        },
                      },
                      "filter": Object {
                        "match_all": Object {},
                      },
                    },
                  },
                  "filter": Object {
                    "bool": Object {
                      "must": Array [],
                    },
                  },
                },
              },
              "from": 0,
              "size": 10,
              "sort": Array [
                Object {
                  "_score": "desc",
                },
              ],
            }
          `)
          return [200, lvl1Mock]
        })

      const response = await runQuery(gql)
      expect(response.body.data).toMatchSnapshot()
      expect(response.status).toEqual(200)
    })

    it('1 lvl1 filter, bring back level 1 + lvl2 facets. Lvl2 facets have lvl1 filter applied', async () => {
      setupTestServer({
        config,
        addToQueryType: true,
        typeName: 'ResultSet',
        hitTypeName: 'ResultHit'
      })

      const gql = `
        {
          results(filters: [{identifier: "categories", value: "Coats and Jackets", level: 1 }]) {
            hits(page: {size: 10, from: 0 }) {
              items {
                id
              }
            }
            facets {
              identifier
              display
              label
              type
              display
              ... on HierarchicalMenuFacet {
                entries {
                  count
                  label
                  entries {
                    count
                    label
                    entries {
                      count
                      label
                    }
                  }
                }
              }
            }
          }
        }
      `

      const scope = nock('http://localhost:9200')
        .post('/movies/_search')
        .reply((uri, body) => {
          expect(body).toMatchInlineSnapshot(`
            Object {
              "_source": Object {
                "includes": Array [
                  "actors",
                  "writers",
                ],
              },
              "aggs": Object {
                "facet_bucket_all": Object {
                  "aggs": Object {},
                  "filter": Object {
                    "bool": Object {
                      "must": Array [
                        Object {
                          "bool": Object {
                            "must": Array [
                              Object {
                                "term": Object {
                                  "category_lvl1.keyword": "Coats and Jackets",
                                },
                              },
                            ],
                          },
                        },
                      ],
                    },
                  },
                },
                "facet_bucket_categories": Object {
                  "aggs": Object {
                    "categories": Object {
                      "aggs": Object {
                        "lvl_1": Object {
                          "aggs": Object {
                            "aggs": Object {
                              "terms": Object {
                                "field": "category_lvl1.keyword",
                              },
                            },
                          },
                          "filter": Object {
                            "match_all": Object {},
                          },
                        },
                        "lvl_2": Object {
                          "aggs": Object {
                            "aggs": Object {
                              "terms": Object {
                                "field": "category_lvl2.keyword",
                              },
                            },
                          },
                          "filter": Object {
                            "bool": Object {
                              "must": Array [
                                Object {
                                  "term": Object {
                                    "category_lvl1.keyword": Object {
                                      "value": "Coats and Jackets",
                                    },
                                  },
                                },
                              ],
                            },
                          },
                        },
                      },
                      "filter": Object {
                        "match_all": Object {},
                      },
                    },
                  },
                  "filter": Object {
                    "bool": Object {
                      "must": Array [],
                    },
                  },
                },
              },
              "from": 0,
              "post_filter": Object {
                "bool": Object {
                  "must": Array [
                    Object {
                      "bool": Object {
                        "must": Array [
                          Object {
                            "term": Object {
                              "category_lvl1.keyword": "Coats and Jackets",
                            },
                          },
                        ],
                      },
                    },
                  ],
                },
              },
              "size": 10,
              "sort": Array [
                Object {
                  "_score": "desc",
                },
              ],
            }
          `)
          return [200, lvl2Mock]
        })

      const response = await runQuery(gql)
      expect(response.body.data).toMatchSnapshot()
      expect(response.status).toEqual(200)
    })

    it('1 lvl1 filter, bring back level 1..3 facets. Lvl2 facets have lvl1 filter applied, lvl3 will have lvl 1..2 filter applied', async () => {
      setupTestServer({
        config,
        addToQueryType: true,
        typeName: 'ResultSet',
        hitTypeName: 'ResultHit'
      })

      const gql = `
        {
          results(filters: [{identifier: "categories", value: "Coats and Jackets", level: 1 }, {identifier: "categories", value: "Leather jackets", level: 2 }]) {
            hits(page: {size: 10, from: 0 }) {
              items {
                id
              }
            }
            facets {
              identifier
              display
              label
              type
              display
              ... on HierarchicalMenuFacet {
                entries {
                  count
                  label
                  entries {
                    count
                    label
                    entries {
                      count
                      label
                    }
                  }
                }
              }
            }
          }
        }
      `

      const scope = nock('http://localhost:9200')
        .post('/movies/_search')
        .reply((uri, body) => {
          expect(body).toMatchInlineSnapshot(`
            Object {
              "_source": Object {
                "includes": Array [
                  "actors",
                  "writers",
                ],
              },
              "aggs": Object {
                "facet_bucket_all": Object {
                  "aggs": Object {},
                  "filter": Object {
                    "bool": Object {
                      "must": Array [
                        Object {
                          "bool": Object {
                            "must": Array [
                              Object {
                                "term": Object {
                                  "category_lvl1.keyword": "Coats and Jackets",
                                },
                              },
                              Object {
                                "term": Object {
                                  "category_lvl2.keyword": "Leather jackets",
                                },
                              },
                            ],
                          },
                        },
                      ],
                    },
                  },
                },
                "facet_bucket_categories": Object {
                  "aggs": Object {
                    "categories": Object {
                      "aggs": Object {
                        "lvl_1": Object {
                          "aggs": Object {
                            "aggs": Object {
                              "terms": Object {
                                "field": "category_lvl1.keyword",
                              },
                            },
                          },
                          "filter": Object {
                            "match_all": Object {},
                          },
                        },
                        "lvl_2": Object {
                          "aggs": Object {
                            "aggs": Object {
                              "terms": Object {
                                "field": "category_lvl2.keyword",
                              },
                            },
                          },
                          "filter": Object {
                            "bool": Object {
                              "must": Array [
                                Object {
                                  "term": Object {
                                    "category_lvl1.keyword": Object {
                                      "value": "Coats and Jackets",
                                    },
                                  },
                                },
                              ],
                            },
                          },
                        },
                        "lvl_3": Object {
                          "aggs": Object {
                            "aggs": Object {
                              "terms": Object {
                                "field": "category_lvl3.keyword",
                              },
                            },
                          },
                          "filter": Object {
                            "bool": Object {
                              "must": Array [
                                Object {
                                  "term": Object {
                                    "category_lvl1.keyword": Object {
                                      "value": "Coats and Jackets",
                                    },
                                  },
                                },
                                Object {
                                  "term": Object {
                                    "category_lvl2.keyword": Object {
                                      "value": "Leather jackets",
                                    },
                                  },
                                },
                              ],
                            },
                          },
                        },
                      },
                      "filter": Object {
                        "match_all": Object {},
                      },
                    },
                  },
                  "filter": Object {
                    "bool": Object {
                      "must": Array [],
                    },
                  },
                },
              },
              "from": 0,
              "post_filter": Object {
                "bool": Object {
                  "must": Array [
                    Object {
                      "bool": Object {
                        "must": Array [
                          Object {
                            "term": Object {
                              "category_lvl1.keyword": "Coats and Jackets",
                            },
                          },
                          Object {
                            "term": Object {
                              "category_lvl2.keyword": "Leather jackets",
                            },
                          },
                        ],
                      },
                    },
                  ],
                },
              },
              "size": 10,
              "sort": Array [
                Object {
                  "_score": "desc",
                },
              ],
            }
          `)
          return [200, lvl3Mock]
        })

      const response = await runQuery(gql)
      expect(response.body.data).toMatchSnapshot()
      expect(response.status).toEqual(200)
    })
  })
})
