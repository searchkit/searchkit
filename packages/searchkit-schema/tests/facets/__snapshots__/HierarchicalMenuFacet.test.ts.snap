// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HierarchicalMenuFacet should return as expected 1 lvl1 filter, bring back level 1 + lvl2 facets. Lvl2 facets have lvl1 filter applied 2`] = `
Object {
  "results": Object {
    "facets": Array [
      Object {
        "display": "HierarchicalMenuFacet",
        "entries": Array [
          Object {
            "count": 426,
            "entries": null,
            "label": "T-Shirts",
          },
          Object {
            "count": 372,
            "entries": null,
            "label": "Casual Shirts",
          },
          Object {
            "count": 324,
            "entries": null,
            "label": "Knitwear",
          },
          Object {
            "count": 308,
            "entries": null,
            "label": "Sweats",
          },
          Object {
            "count": 231,
            "entries": Array [
              Object {
                "count": 107,
                "entries": null,
                "label": "Lightweight & waterproof jackets",
              },
              Object {
                "count": 39,
                "entries": null,
                "label": "Bomber jackets",
              },
              Object {
                "count": 23,
                "entries": null,
                "label": "Leather jackets",
              },
              Object {
                "count": 20,
                "entries": null,
                "label": "Down jackets",
              },
              Object {
                "count": 17,
                "entries": null,
                "label": "Denim jackets",
              },
              Object {
                "count": 14,
                "entries": null,
                "label": "Gilets",
              },
              Object {
                "count": 4,
                "entries": null,
                "label": "Parkas",
              },
              Object {
                "count": 4,
                "entries": null,
                "label": "Raincoats & trench coats",
              },
              Object {
                "count": 3,
                "entries": null,
                "label": "Winter coats",
              },
            ],
            "label": "Coats and Jackets",
          },
          Object {
            "count": 220,
            "entries": null,
            "label": "Polo shirts",
          },
          Object {
            "count": 132,
            "entries": null,
            "label": "Shorts",
          },
          Object {
            "count": 117,
            "entries": null,
            "label": "Trousers",
          },
          Object {
            "count": 113,
            "entries": null,
            "label": "Swimwear",
          },
          Object {
            "count": 80,
            "entries": null,
            "label": "Underwear",
          },
        ],
        "identifier": "categories",
        "label": "Categories",
        "type": "HierarchicalMenuFacet",
      },
    ],
    "hits": Object {
      "items": Array [
        Object {
          "id": "10163292707229244",
        },
        Object {
          "id": "13452677152061433",
        },
        Object {
          "id": "11452292647194357",
        },
        Object {
          "id": "560971904241887",
        },
        Object {
          "id": "10163292707229255",
        },
        Object {
          "id": "13452677152333220",
        },
        Object {
          "id": "11452292647412043",
        },
        Object {
          "id": "9465239339535960",
        },
        Object {
          "id": "13452677150717069",
        },
        Object {
          "id": "3633577411025988",
        },
      ],
    },
  },
}
`;

exports[`HierarchicalMenuFacet should return as expected 1 lvl1 filter, bring back level 1..3 facets. Lvl2 facets have lvl1 filter applied, lvl3 will have lvl 1..2 filter applied 2`] = `
Object {
  "results": Object {
    "facets": Array [
      Object {
        "display": "HierarchicalMenuFacet",
        "entries": Array [
          Object {
            "count": 426,
            "entries": null,
            "label": "T-Shirts",
          },
          Object {
            "count": 372,
            "entries": null,
            "label": "Casual Shirts",
          },
          Object {
            "count": 324,
            "entries": null,
            "label": "Knitwear",
          },
          Object {
            "count": 308,
            "entries": null,
            "label": "Sweats",
          },
          Object {
            "count": 231,
            "entries": Array [
              Object {
                "count": 107,
                "entries": null,
                "label": "Lightweight & waterproof jackets",
              },
              Object {
                "count": 39,
                "entries": null,
                "label": "Bomber jackets",
              },
              Object {
                "count": 23,
                "entries": Array [
                  Object {
                    "count": 107,
                    "label": "Long",
                  },
                  Object {
                    "count": 39,
                    "label": "Short",
                  },
                ],
                "label": "Leather jackets",
              },
              Object {
                "count": 20,
                "entries": null,
                "label": "Down jackets",
              },
              Object {
                "count": 17,
                "entries": null,
                "label": "Denim jackets",
              },
              Object {
                "count": 14,
                "entries": null,
                "label": "Gilets",
              },
              Object {
                "count": 4,
                "entries": null,
                "label": "Parkas",
              },
              Object {
                "count": 4,
                "entries": null,
                "label": "Raincoats & trench coats",
              },
              Object {
                "count": 3,
                "entries": null,
                "label": "Winter coats",
              },
            ],
            "label": "Coats and Jackets",
          },
          Object {
            "count": 220,
            "entries": null,
            "label": "Polo shirts",
          },
          Object {
            "count": 132,
            "entries": null,
            "label": "Shorts",
          },
          Object {
            "count": 117,
            "entries": null,
            "label": "Trousers",
          },
          Object {
            "count": 113,
            "entries": null,
            "label": "Swimwear",
          },
          Object {
            "count": 80,
            "entries": null,
            "label": "Underwear",
          },
        ],
        "identifier": "categories",
        "label": "Categories",
        "type": "HierarchicalMenuFacet",
      },
    ],
    "hits": Object {
      "items": Array [
        Object {
          "id": "10163292707229244",
        },
        Object {
          "id": "13452677152061433",
        },
        Object {
          "id": "11452292647194357",
        },
        Object {
          "id": "560971904241887",
        },
        Object {
          "id": "10163292707229255",
        },
        Object {
          "id": "13452677152333220",
        },
        Object {
          "id": "11452292647412043",
        },
        Object {
          "id": "9465239339535960",
        },
        Object {
          "id": "13452677150717069",
        },
        Object {
          "id": "3633577411025988",
        },
      ],
    },
  },
}
`;

exports[`HierarchicalMenuFacet should return as expected No filters, bring back only level 1 facets 2`] = `
Object {
  "results": Object {
    "facets": Array [
      Object {
        "display": "HierarchicalMenuFacet",
        "entries": Array [
          Object {
            "count": 426,
            "entries": null,
            "label": "T-Shirts",
          },
          Object {
            "count": 372,
            "entries": null,
            "label": "Casual Shirts",
          },
          Object {
            "count": 324,
            "entries": null,
            "label": "Knitwear",
          },
          Object {
            "count": 308,
            "entries": null,
            "label": "Sweats",
          },
          Object {
            "count": 231,
            "entries": null,
            "label": "Coats and Jackets",
          },
          Object {
            "count": 220,
            "entries": null,
            "label": "Polo shirts",
          },
          Object {
            "count": 132,
            "entries": null,
            "label": "Shorts",
          },
          Object {
            "count": 117,
            "entries": null,
            "label": "Trousers",
          },
          Object {
            "count": 113,
            "entries": null,
            "label": "Swimwear",
          },
          Object {
            "count": 80,
            "entries": null,
            "label": "Underwear",
          },
        ],
        "identifier": "categories",
        "label": "Categories",
        "type": "HierarchicalMenuFacet",
      },
    ],
    "hits": Object {
      "items": Array [
        Object {
          "id": "10163292707229244",
        },
        Object {
          "id": "13452677152061433",
        },
        Object {
          "id": "11452292647194357",
        },
        Object {
          "id": "560971904241887",
        },
        Object {
          "id": "10163292707229255",
        },
        Object {
          "id": "13452677152333220",
        },
        Object {
          "id": "11452292647412043",
        },
        Object {
          "id": "9465239339535960",
        },
        Object {
          "id": "13452677150717069",
        },
        Object {
          "id": "3633577411025988",
        },
      ],
    },
  },
}
`;
